<div style="background-color: #f0f7ff;
            border-left: 5px solid #1e90ff;
            border-radius: 6px;
            padding: 24px 20px;
            margin: 20px 0;
            box-sizing: border-box;
            box-shadow: 0 2px 5px rgba(0,0,0,0.05);
            text-align: center;">
    <h1 style="text-align: center; margin: 0 0 10px 0; color: #1f4e79;">Scientific Programming in Python</h1>
    <h3 style="text-align: center; margin: 0 0 5px 0; font-weight: normal; line-height: 1.4;">Karl N. Kirschner</h3>
    <p style="text-align: center; margin: 0; font-size: small; color: #555;">
        Bonn-Rhein-Sieg University of Applied Sciences<br>Sankt Augustin, Germany
    </p>
    <hr style="border: 0; border-top: 3px solid #ccc; margin: 20px auto; width: 100%;">
    <h2 style="text-align: center; margin: 10px 0 0 0; color: #2c3e50;">Significant Figures</h2>
</div>

<p><img alt="Accuracy vs Precision" width="800" src="00_images/alexander-muzenhardt-753371-unsplash.jpg" align="center" hspace="10px" vspace="0px"></p>

## Accuracy versus Precision
### What is the difference between the two?

#### A visual attempt to place the idea of significant figures into context.

1. Plot a target data point and some measured data
   - First using a **large scale** for the **axes**
2. <font color="red">Red</font> = target reference value
3. <font color="dodgerblue">Blue</font> and <font color="orange">Orange</font> = **two** different data sets, measured by **two** different people

<div style="display: flex; gap: 20px; align-items: flex-start; margin-top: 15px;">
  <img src="00_images/accuracy_vs_precision_1.png" alt="Accuracy vs Precision" style="width: 400px; flex-shrink: 0;">
  <div>
    <p>All data points look close together, and thus they all appear to be equally correct.</p>
    <p>However, the <strong>choice</strong> of the <font color="MediumVioletRed"><strong>scale</strong></font> used is <strong>important</strong>.</p>
  </div>
</div>

<p style="font-size: 1.1em; margin-top: 15px;">
  The choice of <strong>how</strong> we choose to <strong>present</strong> the data's <font color="MediumVioletRed"><strong>numbers</strong></font> is <strong>important</strong>.
</p>

<br><br><br>
<br><br><br>
<br><br><br>

<div style="display: flex; gap: 24px; align-items: flex-start; margin-top: 16px;">
  <img src="00_images/accuracy_vs_precision_2.png" alt="Accuracy vs Precision" style="width: 400px; flex-shrink: 0; border-radius: 4px;">
  <div>
    <p></p>
    <p>On a more <strong>appropriate scale</strong> - the data's clustering looks a bit different now.</p>
    <p>Notice how we <strong>added a decimal point</strong> to the axes' <strong>tick numbers</strong>.</p>
    <p>The most <font color="dodgerblue"><strong>accurate</strong></font> value is a single data point (one of the <font color="dodgerblue"><strong>blue</strong></font> points), while the higher <font color="orange"><strong>precision</strong></font> values are represented by the <font color="orange"><strong>orange</strong></font> points.</p>
    <ul style="margin-top: 10px; padding-left: 20px;">
      <li>The person generating the <font color="dodgerblue"><strong>blue</strong></font> data points, was <strong>less consistent</strong> in doing so.</li>
      <li>The person generating the <font color="orange"><strong>orange</strong></font> data points, was <strong>highly consistent</strong> in doing so.</li>
    </ul>
  </div>
</div>

<br><br><br>
<br><br><br>
<br><br><br>

<div style="display: flex; gap: 24px; align-items: flex-start; margin-top: 16px;">
  <img src="00_images/accuracy_vs_precision_3.png" alt="Accuracy vs Precision" style="width: 400px; flex-shrink: 0; border-radius: 4px;">
  <div>
    <p></p>
    <p><strong>Two more decimal points</strong> to the axes' <strong>tick numbers</strong>.</p>
    <p>Notice that when "zooming" into the data, the highly-precise <font color="orange"><strong>orange</strong></font> data also have a distribution.</p>
  </div>
</div>

<div style="margin-top: 20px;">
  <h3 style="margin-bottom: 8px; style="font-size: 14pt;">Important:</h3>
  <p style="font-size: 1.1em;">
    <font size="4">Notice how <font color="MediumVioletRed"><strong>using additional digits</strong></font> to report the tick labels is <font color="MediumVioletRed"><strong>providing</strong></font> us with <font color="MediumVioletRed"><strong>more knowledge</strong></font> about the data that we obtained.</font>
  </p>
</div>

<br><br><br>
<br><br><br>
<br><br><br>

### Accuracy vs. Precision - Data Plot Explanation

<figure style="margin-top: 16px; margin-bottom: 20px;">
  <img src="00_images/accuracy_and_precision.png" alt="Accuracy and Precision" style="width: 600px; max-width: 100%; border-radius: 4px;">
  <figcaption style="font-size: 0.9em; color: #555; margin-top: 6px;">
    (Figure modified from <a href="https://en.wikipedia.org/wiki/Accuracy_and_precision#/media/File:Accuracy_and_precision.svg">Wikipedia</a>)
  </figcaption>
</figure>

<p>
  <strong>Precision (<font color="MediumVioletRed">reproducibility</font>)</strong>: Data point distributions are associated with a range (i.e., an error <font color="Orange"><strong>$\pm$ x.xx</strong></font>) that says something about the <strong>error within their measurement</strong>.
</p>

<p style="margin-top: 10px;">
  One can report a data's <strong>average</strong> value with a <strong>standard deviation</strong> (i.e., <font color="MediumVioletRed"><strong>X.XX $\pm$ x.xx</strong></font>), which will depend on how much "information" the numbers contain.
</p>

<div style="margin-top: 16px;">
  <p><strong>Accuracy (<font color="MediumVioletRed">correctness</font>)</strong>: The <strong>difference</strong> between the <strong>reference value</strong> with the <strong>average measured/generated data</strong></p>
  <ul style="margin-top: 6px; padding-left: 20px;">
    <li>Accuracy Examples
      <ul style="padding-left: 20px;">
        <li>$\color{red}{X.XX \pm x.xx} - \color{dodgerBlue}{X.XX \pm x.xx}$, or</li>
        <li>$\color{red}{X.XX \pm x.xx} - \color{Orange}{X.XX \pm x.xx}$</li>
      </ul>
    </li>
  </ul>
</div>

<br><br><br>

### Take Home Message

1. **Accuracy** is how close the data is to a reference value (i.e., <font color="MediumVioletRed">**error of the data to an ideal value**</font>).
2. **Precision** is how closely repeated measurements agree with each other (i.e., its **distribution**/<font color="Orange">**variance**</font>; the <font color="Orange">**error within the data itself**</font>).
   - Sigfigs are all about precision: the **number of digits** that are **reliably known** from the measurement.
3. The <font color="MediumVioletRed">**number of digits**</font> representing the data is <font color="MediumVioletRed">related</font> to its <font color="MediumVioletRed">**information content**</font> (i.e., <font color="MediumVioletRed">**decimal places = information**</font>).

<div style="background-color: #f0f7ff;
            border-left: 5px solid dodgerblue;
            border-radius: 6px;
            padding: 20px;
            max-width: 650px;
            margin: 30px auto;
            text-align: center;
            box-shadow: 0 2px 5px rgba(0,0,0,0.05)">
    
  <p style="font-size: 1.2em; color: #1e3a8a; margin: 0; line-height: 1.5;">
    <strong>Significant figures</strong> are the way that scientists <strong>communicate</strong> the <strong>uncertainty</strong><br>
    — and thus, the <strong>precision</strong> — of reported <strong>numbers</strong>.
  </p>
</div>


<br><br><br>
<br><br><br>
<br><br><br>

<hr style="border:2px solid gray"></hr>

## Example

<div style="display: flex; gap: 24px; align-items: flex-start; margin-top: 16px;">
  <figure style="margin: 0; flex-shrink: 0;">
    <img src="00_images/satealite_measure_google_ai.png" alt="sigfig_comic" style="width: 450px; border-radius: 4px;">
    <figcaption style="font-size: 0.85em; color: #666; margin-top: 6px;">
      Source: Google AI-generated image (2026).
    </figcaption>
  </figure>

  <div>
    <p>Three researchers are measuring the size of a device that will be inserted into a satellite. This is done independently from one another.</p>
    <ul style="padding-left: 20px; margin: 12px 0;">
      <li><strong>Researcher 1:</strong> 10.3 cm</li>
      <li><strong>Researcher 2:</strong> 10.39 cm</li>
      <li><strong>Researcher 3:</strong> 10.370 cm</li>
    </ul>
    <p>To give their boss their best measurement value, they take the <strong>average of their values</strong>.</p>
    <p style="font-size: 1.1em; margin-top: 16px; color: #1e3a8a;">
      <strong>What is the average value that they should report?</strong>
    </p>
  </div>
</div>

<br><br><br>
<br><br><br>
<br><br><br>

**Possible** ways to answer this:

- Do the math directly

### Solutions Walkthrough

#### 1. Calculate the Unrounded Mean
$$\text{Average} = \frac{10.3 + 10.39 + 10.370}{3} = \frac{31.060}{3} = 10.353333... \text{ cm}$$

In [1]:
(10.3 + 10.39 + 10.370)/3

10.353333333333333

<br><br><br>
<br><br><br>
<br><br><br>

#### 2. Round the Number
$$\text{Average} = \frac{10.3 + 10.39 + 10.370}{3} = \frac{31.060}{3} = 10.353333... \text{ cm}$$
$$\text{Round to 3 decimals} = 10.353 \text{ cm}$$

In [2]:
round(number=(10.3 + 10.39 + 10.370)/3, ndigits=3)

10.353

Or if we do it to represent that the last 3 goes on forever:

$\Large\frac{10.3 + 10.39 + 10.370}{3} = 10.35\bar{3}$


<br><br><br>
<br><br><br>
<br><br><br>

#### 3. **Pre-rounding** of each data point **before** the averaging (i.e., getting the numbers into the **same magnitude**)
  <p style="margin: 0; font-size: 1.1em; text-align: center;">
    $$\frac{\mathbf{10.3} + \mathbf{10.4} + \mathbf{10.4}}{3} = 10.367 \text{ cm}$$
  </p>

In [3]:
round(number=(round(10.3, ndigits=1)
              + round(10.39, ndigits=1)
              + round(10.370, ndigits=1))/3, ndigits=3)

10.367

All of this seems correct, right?

<br><br><br>
<br><br><br>
<br><br><br>

<div style="background-color: #f0f7ff;
            border-left: 5px solid dodgerblue;
            border-radius: 6px;
            padding: 20px;
            max-width: 650px;
            margin: 30px auto;
            text-align: left;
            box-shadow: 0 2px 5px rgba(0,0,0,0.05)">

  <h3 style="color: MediumVioletRed; margin-top: 0; margin-bottom: 12px;">
    Why All Three Approaches Are Incorrect
  </h3>

  <p><strong>1. Unrounded Raw Value (10.353333... cm)</strong><br>
  Implies infinite precision. It claims we know the device's size down to sub-nanometers - none of the researchers could do this.</p>

  <p><strong>2. Arbitrary Decimal Rounding (10.353 cm)</strong><br>
  Picking 3 decimal places is an arbitrary choice here - it ignores the actual measurement uncertainty.</p>

  <p><strong>3. Pre-Rounding Input Data (10.367 cm)</strong><br>
  Throws away valid, high-precision measurement data (Researchers 2 and 3) <em>before</em> doing the math - introduces rounding error into the intermediate sum.</p>
</div>

<br><br><br>
<br><br><br>
<br><br><br>

The **correct answer** is: <font color='MediumVioletRed'>**10.4**</font>

<!-- Data set 2 (i.e.10.36, 10.3, 10.35): 10.3 -->

**Why report the number as <font color='MediumVioletRed'>XX.Y</font>?**


<br><br><br>
<br><br><br>
<br><br><br>

<hr style="border:2px solid gray"></hr>

## Significant Figures / Significant Digits / Sigfigs

**What are sigfigs?**

<font color="dodgerblue"><strong>Sigfigs usage reveals the information content - the precision - of a number.</strong></font>


## Firm Rules for Sigfigs

<div style="background-color: #f8f9fa; border-left: 4px solid dodgerblue; padding: 14px 18px; border-radius: 4px; margin-bottom: 12px;">
  <p style="margin: 0;">
    <strong>1. All non-zero digits are significant:</strong> <font color="dodgerblue"><strong>1, 2, 3, 4, 5, 6, 7, 8, 9</strong></font>
  </p>
</div>

<div style="background-color: #f8f9fa; border-left: 4px solid dodgerblue; padding: 14px 18px; border-radius: 4px; margin-bottom: 12px;">
  <p style="margin: 0 0 8px 0;">
    <strong>2. Zeros <font color="dodgerblue">between</font> non-zero digits are significant:</strong> 1<font color="dodgerblue"><strong>0</strong></font>2.0, 2<font color="dodgerblue"><strong>00</strong></font>5.0, 5<font color="dodgerblue"><strong>000</strong></font>9.0
  </p>
  <p style="margin: 0; font-size: 0.95em; color: #555;">
    <em>Note:</em> The total number of sigfigs in a number like <font color="red">101</font> (without an explicit decimal point) is technically <font color="red">unclear</font>. We will revisit this below.
  </p>
</div>

<div style="background-color: #fff8f8; border-left: 4px solid dodgerblue; padding: 14px 18px; border-radius: 4px; margin-bottom: 12px;">
  <p style="margin: 0 0 6px 0;">
    <strong>3. Leading zeros are <font color="red">never</font> significant:</strong> <font color="red">00</font>1, <font color="red">0.0</font>2, <font color="red">00</font>1.887, <font color="red">0.000</font>515
  </p>
  <ul style="margin: 0; padding-left: 20px; line-height: 1.6;">
    <li>They are <strong>placeholders</strong> that position the decimal point.</li>
    <li>In <strong>scientific notation</strong>, they disappear $\rightarrow$ $0.000515$ becomes $5.15 \times 10^{-4}$.</li>
  </ul>
</div>

<div style="background-color: #f8f9fa; border-left: 4px solid dodgerblue; padding: 14px 18px; border-radius: 4px; margin-bottom: 16px;">
  <p style="margin: 0;">
    <strong>4. Trailing zeros after a decimal place are significant:</strong> 0.02<font color="dodgerblue"><strong>0</strong></font>
  </p>
</div>

---

<div style="background-color: #fff8f0; border: 1px solid dodgerblue; border-radius: 6px; padding: 14px 18px; margin-top: 16px;">
  <p style="margin: 0 0 6px 0; color: #9a3412;">
    <strong>Side Note on Scientific Notation Syntax:</strong>
  </p>
  <p style="margin: 0 0 6px 0;">
    Scientific notation can be represented in any of the following equivalent forms:
  </p>
  <ul style="margin: 0; padding-left: 20px; line-height: 1.6;">
    <li>$5.15 \times 10^{-4}$</li>
    <li><code>5.15E-4</code></li>
    <li><code>5.15e-4</code></li>
  </ul>
  <p style="margin: 8px 0 0 0; font-size: 0.85em; color: #78350f;">
    <em>Source:</em> <a href="https://en.wikipedia.org/wiki/Scientific_notation" target="_blank">Wikipedia — Scientific Notation</a>
  </p>
</div>

## Additional Focus: Surrounding Zeros and the Use of Decimals

Numbers with zeros can be tricky. Always ask yourself:  
> *"Is the zero acting as a **placeholder**, or is it a **measured digit**?"*

---

<div style="background-color: #f8f9fa; border-left: 4px solid dodgerblue; padding: 14px 18px; border-radius: 4px; margin-bottom: 12px;">
  <p style="margin: 0;">
    <strong>1. Zeros after a decimal (e.g., 120.00)</strong><br>
    <font color="dodgerblue"><strong>Always significant</strong></font> if preceded by a significant digit.<br>
    <em>Contrast:</em> <code>0.01</code> has only <strong>1 sigfig</strong> because leading zeros are solely placeholders.
  </p>
</div>

<div style="background-color: #f8f9fa; border-left: 4px solid dodgerblue; padding: 14px 18px; border-radius: 4px; margin-bottom: 12px;">
  <p style="margin: 0;">
    <strong>2. Zeros before an explicit decimal (e.g., 120.)</strong><br>
    <font color="dodgerblue"><strong>Significant</strong></font> — <code>120.</code> has <strong>3 sigfigs</strong>.<br>
    The trailing decimal explicitly specifies that all digits to the left were measured (no rounding occurred).
  </p>
</div>

<div style="background-color: #fff8f8; border-left: 4px solid dodgerblue; padding: 14px 18px; border-radius: 4px; margin-bottom: 20px;">
  <p style="margin: 0;">
    <strong>3. Zeros without a decimal (e.g., 120)</strong><br>
    <font color="red"><strong>Not significant</strong></font> — <code>120</code> indicates <strong>2 sigfigs</strong>.<br>
    The absence of a decimal creates ambiguity regarding measurement precision.
  </p>
</div>

---

### Case Study: How Precise is "1000 grams"?

Without a <strong>decimal point</strong> or additional <strong>context</strong>, evaluating <code>1000 g</code> forces us to make assumptions:
- *Where did this value come from? Is it an average, a sum, or a single measurement?*
- *Is it a defined exact unit conversion (1000 g = 1 kg) or a physical measurement?*

While people often **assume** numbers without a decimal are precise to the ones place, the underlying true value could easily be `1000.1`, `1000.0`, or `999.8`.

#### Demonstration of How "1000" Might Be Obtained:

<div style="background-color: #f8f9fa; border: 1px solid #e2e8f0; border-radius: 6px; padding: 16px; margin-top: 12px;">
  <ul style="margin: 0; padding-left: 20px; line-height: 1.7;">
    <li>
      $920 + 80 = 1000$  
      <span style="color: #64748b;">(Ambiguous and imprecise — how many zeros were actually measured?)</span>
    </li>
    <li>
      $919.7 + 80.0 = 999.7 \approx 1000$  
      <span style="color: #64748b;">(Rounding $999.7 \rightarrow 1000$ loses significant information about number's certainty)</span>
    </li>
    <li>
      $920. + 80. = 1000.$  
      <span style="color: #64748b;">(Explicit decimal indicates precision down to the <strong>ones place</strong>)</span>
    </li>
    <li>
      $920.0 + 80.0 = 1000.0$  
      <span style="color: #64748b;">(Explicit decimal indicates precision down to the <strong>tenths place</strong>)</span>
    </li>
  </ul>
</div>

### Examples of Numbers and Their Sigfigs
    
| Example | Comparison | Sigfigs |
|:---|:---:|---:|
| 1 | $1.23 \times 10^2$ <br> <font color='dodgerblue'>vs.</font> <br> $110.$ | **3**  <br><br>  **3** |
| | | |
| 2 | $123.00$ <br> <font color='dodgerblue'>vs.</font> <br> $0.0012300$ | **5** <br><br>  **5** |
| | | |
| 3 | $100$ <br> <font color='dodgerblue'>vs.</font> <br> $100.$ (i.e., $1.00 \times 10^2$) | **unclear**/**inexact** <br><br> **3** |
| | | |
| 4 | $120$ <br> <font color='dodgerblue'>vs.</font> <br> $1.2 \times 10^2$ | **unclear**/**inexact** <br><br> **2** |

<hr style="border:2px solid gray"></hr>

## Sigfigs and Math

<b>Addition and Subtraction</b> - requires that the answer's number has places <font color='dodgerblue'>after the <b>decimal</b> that is equal to the <b>least</b> number of <b>decimals</b> present <b>in the calculation.</b></font>

- This rule is about <font color='dodgerblue'><b>absolute uncertainty</b> (i.e., the <b>number of decimal places</b>)</font>.
    - For illustration: If one of your numbers is only known to the tenths place, then the sum/difference is also only reliable to the tenths place — you can’t justify reporting hundredths place in the final answer.

Examples:
<ol type = "1">
    <li> 7.<font color='dodgerblue'>1</font> + 2.<font color='dodgerblue'>3</font> = 9.<font color='dodgerblue'>4</font><br><br>
    <li> 0.<font color='dodgerblue'>003</font> + 0.<font color='dodgerblue'>020</font> = 0.<font color='dodgerblue'>023</font><br><br>
    <li> 15.<font color='dodgerblue'>0</font> + 0.55 = 15.<font color='dodgerblue'>6</font><br>Unrounded value = <font color='IndianRed'>15.55</font><br><br>
    <li> 10.<font color='dodgerblue'>4</font> + 10. = 2.<font color='dodgerblue'>0</font>e1 = 20.<br><b>Key</b>: Convert to <b>scientific notation:</b> 1.04e1 + 1.0e1 = 2.04e1 $\rightarrow$ 2.0e1<br><br>
    <li> 8.83210<b>e3</b> - <font color='dodgerblue'>8.32<b>e2</b></font> = 8.<font color='dodgerblue'>000</font>e3<br>Convert values to have the <b>same exponential form</b> and then compute:<br>8.83210<b>e3</b> - 0.832<b>e3</b> = (8.83210 - 0.<font color='dodgerblue'>832</font>)e3 $\rightarrow$ 8000.1e3 = 8.<font color='dodgerblue'>000</font>e3 <br> or <br>88.3210<b>e2</b> - 8.32<b>e2</b> = (88.3210 - 8.<font color='dodgerblue'>32</font>)e2 $\rightarrow$ 80.00e2 = 8.<font color='dodgerblue'>000</font>e3<br><br>
    <li> 0.003 + 0.<font color='dodgerblue'>02</font> = 0.<font color='dodgerblue'>02</font><br>There is less "information" in 0.02 in comparison to 0.020. Thus, the <b>more precise 0.003</b> is lost within the <b>less precise 0.02</b>.<br><br>

</ol>

<b>Multiplication and Division</b> - requires that the answer's number has significant figure places <font color='dodgerblue'>equal to the number of the least amount of significant figures used <b>in the calculation</b>.</font>

- This rule is about <font color='dodgerblue'><b>relative uncertainty</b> (i.e., <b>total number of sigfigs</b>)</font>.
    - When multiplying/dividing, the errors (uncertainties) in the input values are multiplied/divided as well. The results cannot be relatively more precise than the least relatively precise input.

Examples:
<ol type = "1">
    <li> <font color='dodgerblue'>8.21</font> x <font color='dodgerblue'>2.32</font>   = <font color='dodgerblue'>19.0</font> (Unrounded value = <font color='IndianRed'>19.0472</font>)<br><br>
    <li> 8.210 x <font color='dodgerblue'>2.32</font>  = <font color='dodgerblue'>19.0</font><br><br>
    <li> 0.<font color='dodgerblue'>21</font> x 10.0 = <font color='dodgerblue'>2.1</font><br><br>
    <li> <font color='dodgerblue'>123</font>0 x 2.000 = <font color='dodgerblue'>2.46</font>e3<br>Uncertain value of 1230, but maybe one could say that at a minimum three sigfigs -> the 1, 2 and 3 (but still not good). This illustrates the importance of carefully specifying numbers with regard to sigfigs.<br><br>
    <li> <font color='dodgerblue'>1230.</font> x <font color='dodgerblue'>2.000</font> = <font color='dodgerblue'>2460.</font> or <font color='dodgerblue'>2.460e3</font><br><br>
    <li> <font color='dodgerblue'>8.2</font> x 12.3 = <font color='dodgerblue'>1.0e2</font><br>Unrounded value = <font color='IndianRed'>100.86</font> - we <b>must</b> use scientific notation to give the results with the correct number of sigfigs (i.e., <font color='IndianRed'>100.</font> has 3 sigfigs; <font color='IndianRed'>100</font> is unclear).<br><br>
    <li> <font color='dodgerblue'>2.0</font> x 0.625 = <font color='dodgerblue'>1.3</font><br>Unrounded value = <font color='IndianRed'>1.25</font>
</ol>


<div style="background-color: #fff8f0;
            border-left: 5px solid dodgerblue;
            border-radius: 6px;
            padding: 20px;
            max-width: 650px;
            margin: 30px auto;
            text-align: left;
            box-shadow: 0 2px 5px rgba(0,0,0,0.05)">
  <p style="margin: 0 0 8px 0; color: #1e293b;">
    <strong>The Golden Rule:</strong>
  </p>
  <ul style="margin: 0; padding-left: 20px;">
    <li><strong></strong>Always carry <strong><em>full precision</em></strong> through intermediate calculations (i.e., the math steps), <strong>applying sigfig rules only to the final result</strong>.</li>
  </ul>
</div>

<hr style="border:2px solid gray"></hr>

## Numbers with Infinite Sigfigs


**Exact Number**: has an <font color='dodgerblue'>infinite number of significant figures</font> — must have some context given

<ol type = "1">
<li> Numbers that define other numbers. For example, 100 cm in 1 m — both numbers are exact
<li> Whole quantity that can not exist in a fraction. For example, 2 molecules; 8 planets and 146 moons in our solar system
<li> Mathematically derived integers in equations. For example, taking the average of $5.0 + 5.0 = \frac{5.0+5.0}{2}$ — the <b>2</b> is <b>exact</b>, while 5.0 has two sigfigs (thus, the limiting value)
</ol>

**Mathematical Constants**: are also <font color='dodgerblue'>infinite</font> (but might be given as approximate value)
<ol type = "1">    
<li> c (speed of light in a vacuum): 299 792 458 m s$^{-1}$ (exact) (i.e., $299 792 458.\bar0$)
<li> $\pi$ (Pi): ca. 3.14159 26535 89793 23846 ...
<li> $e$ (Euler's constant): ca. 2.71828 18284 59045 23536 ...
<li> $\sqrt{2}$ (Pythagoras's constant): ca. 1.41421 35623 73095 04880 ...   
</ol>

https://physics.nist.gov/cuu/Constants

**Additional Resources**

https://www.khanacademy.org/math/arithmetic-home/arith-review-decimals/arithmetic-significant-figures-tutorial/v/significant-figures

<hr style="border:2px solid gray"></hr>

#### A little bit of number humor
<figure>
  <img alt="sigfig_comic" width="500" src="00_images/xkcd_large_number_formats-2.png" hspace="10px" vspace="10px">
    
    Source: XKCD (https://xkcd.com/2319).
</figure>